tensorflow/tools/compatibility/ipynb.py

```diff
@@ -0,0 +1,138 @@
+# Copyright 2019 The TensorFlow Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# =============================================================================
+
+"""A module to support operation on ipynb files"""
+
+from __future__ import print_function
+
+import collections
+import copy
+import json
+import shutil
+import tempfile
+
+
+from tensorflow.tools.compatibility import ast_edits
+from tensorflow.tools.compatibility import tf_upgrade_v2
+
+
+CodeLine = collections.namedtuple('CodeLine', ['cell_number', 'code'])
+
+
+def process_file(in_filename, out_filename):
+  """The function where we inject the support for ipynb upgrade."""
+  upgrader = ast_edits.ASTCodeUpgrader(tf_upgrade_v2.TFAPIChangeSpec())
+
+  if in_filename.endswith('.py'):
+    files_processed, report_text, errors = \
+      upgrader.process_file(in_filename, out_filename)
+
+  elif in_filename.endswith('.ipynb'):
```

Comment:
Can you move the dispatch logic to the main function (and possibly maybe even process_tree)